### DISCUSSION TITLE
Getting "fallible argument" with handled error

### DISCUSSION THREAD

**Discussion by ElToro1966** [vectordotdev/vector#21767] [Q&A]
I am using Vector to move data from a Kafka broker to Clickhouse. In the transforms in vector.yaml, I've got:

```

      transaction_time, err = parse_timestamp(.transaction_time, format: "%Y-%m-%d %H:%M:%S.%6f")

      if err != null {

        log("could not parse timestamp" + .transaction_time + " using format: " + err)

      } else {

        .transaction_time = transaction_time

      }

```

When I run Vector, I get:

```
2024-11-12T09:04:54.506894Z ERROR vector::topology::builder: Configuration error. error=Transform "field_mapping": 

error[E630]: fallible argument

  ┌─ :8:7

  │

8 │   log("could not parse timestamp" + .transaction_time + " using format: " + err)

  │       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  │       │

  │       this expression can fail

  │       handle the error before passing it in as an argument
```

What am I not seeing? I have been following https://vector.dev/docs/reference/vrl/errors/#handling, so this should work...

Thanks

**Comment by jszwedko** [comment]
Hi @ElToro1966 ,

Here I think the issue is that the VRL compiler isn't smart enough to realize that `err` will definitely be a string and not, possibly, `null`. Since you know it is a string due to the `if` condition preceding it, I would change `log` to `log!` since it will never fail in practice.

In the future, the compiler could get smart enough to realize that all of the arguments in the string you are building are strings themselves (and include proper string templating in VRL).

Hope this helps!

**Reply by ElToro1966** [reply]
Thank you for the suggestion, @jszwedko . I still get "fallible argument" with "log!", but your comment got me on the right track. This works:
`log("could not parse timestamp" + to_string!(.transaction_time) + " using format: " + err)`

**Reply by jszwedko** [reply]
Aha, yeah, I see, because the error is during the concatenation via the `+` sign. Glad you got it working!
